packages/taiko-client/driver/preconf_blocks/server.go

```diff
@@ -313,6 +397,279 @@ func (s *PreconfBlockAPIServer) OnUnsafeL2Payload(
 		return fmt.Errorf("failed to try importing child blocks from cache: %w", err)
 	}
 
+	if msg.EndOfSequencing != nil && *msg.EndOfSequencing && s.rpc.L1Beacon != nil {
+		s.sequencingEndedForEpoch.Add(s.rpc.L1Beacon.CurrentEpoch(), struct{}{})
+
+		notification := map[string]interface{}{
+			"currentEpoch":    s.rpc.L1Beacon.CurrentEpoch(),
+			"endOfSequencing": true,
+		}
+
+		s.wsMutex.Lock()
+		for conn := range s.wsClients {
+			if err := conn.WriteJSON(notification); err != nil {
+				conn.Close()
+				delete(s.wsClients, conn)
+			}
+		}
+		s.wsMutex.Unlock()
+	}
+
+	return nil
+}
+
+// OnUnsafeL2Response implements the p2p.GossipIn interface.
+func (s *PreconfBlockAPIServer) OnUnsafeL2Response(
+	ctx context.Context,
+	from peer.ID,
+	msg *eth.ExecutionPayloadEnvelope,
+) error {
```

Comment:
Shall we also check and call `s.sequencingEndedForEpoch.Add` in `OnUnsafeL2Response`?

Reply:
as you can only query a parent block (because you only know if you missed a block when another block comes in ,and the parent isnt in the cache or canonical chain), you will never receive an `OnUnsafeL2Response` with an EndOfSequencing marker. If you miss the *final block*, you missed it, and have already begun sequencing (if you are the sequencer) on top of the last block you have received, after the handover_ms_buffer of the sequencer.

Reply:
I implemented the above feature.